fuel-core/src/query.rs

```diff
@@ -0,0 +1,183 @@
+    /// Get the signature of a fuel block.
+    fn signature(
+        &self,
+        block_id: &Bytes32,
+    ) -> Result<Option<fuel_crypto::Signature>, KvStoreError>;
+    /// Get the fuel block.
+    fn block(&self, block_id: &Bytes32) -> Result<Option<FuelBlockDb>, KvStoreError>;
+}
+
+/// Generate an output proof.
+pub async fn output_proof(
+    data: &(dyn OutputProofData + Send + Sync),
+    transaction_id: Bytes32,
+    message_id: MessageId,
+) -> Result<Option<OutputProof>, KvStoreError> {
+    // Check if the receipts for this transaction actually contain this message id or exit.
+    if !data.receipts(&transaction_id)?.into_iter().any(
+        |r| matches!(r, Receipt::MessageOut { message_id: id, .. } if id == message_id),
+    ) {
+        return Ok(None)
+    }
+
+    // Get the block id from the transaction status if it's ready.
+    let block_id = data
+        .transaction_status(&transaction_id)?
+        .and_then(|status| match status {
+            TransactionStatus::Failed { block_id, .. }
+            | TransactionStatus::Success { block_id, .. } => Some(block_id),
+            TransactionStatus::Submitted { .. } => None,
+        });
+
+    // Exit if the status doesn't exist or is not ready.
+    let block_id = match block_id {
+        Some(b) => b,
+        None => return Ok(None),
+    };
+
+    // Track if we have found the message we are looking for.
+    let mut message_found = false;
+
+    // Get the message ids in the same order as the transactions.
+    let leaves = data
+        .transactions_on_block(&block_id)?
+        .into_iter()
+        // Filter out transactions that contain no messages 
+        // and get the receipts for the rest.
+        .filter_map(|transaction_id| match data.transaction(&transaction_id) {
+            Ok(transaction) => transaction?
+                .outputs()
+                .iter()
+                .any(Output::is_message)
+                .then(|| data.receipts(&transaction_id)),
+            Err(e) => Some(Err(e)),
+        })
+        // Flatten the receipts after filtering on output messages
+        // and mapping to message ids.
+        .flat_map(|receipts| match receipts {
+            Ok(receipts) => {
+                let iter: Box<dyn Iterator<Item = _>> =
+                    Box::new(receipts.into_iter().filter_map(|r| match r {
+                        Receipt::MessageOut { message_id, .. } => Some(Ok(message_id)),
+                        _ => None,
+                    }));
+                iter
+            }
+            Err(e) => {
+                // Boxing is required because of the different iterator types
+                // returned depending on the error case.
+                let iter: Box<dyn Iterator<Item = _>> = Box::new(std::iter::once(Err(e)));
+                iter
+            }
+        })
+        // Take all message ids up to and including the
+        // requested id.
```

Comment:
Wouldn't ending the iteration here cause this merkle tree to have a different root than the block? We need to add all message ids in the block to the tree to regenerate the same root as the block header, and then get a proof to the matching root.